src/main/java/org/folio/eusage/reports/api/EusageReportsApi.java

```diff
@@ -222,60 +239,167 @@ private void endStream(RowStream<Row> stream, RoutingContext ctx,
+        + " WHERE matchCriteria = $1")
+        .execute(Tuple.of(match))
         .compose(res1 -> {
           if (res1.iterator().hasNext()) {
-            return Future.succeededFuture(res1.iterator().next().getUUID(0));
+            Row row = res1.iterator().next();
+            UUID id = row.getUUID(0);
+            if (row.getString(1) != null) {  // matched with counter reports already?
+              return Future.succeededFuture(id);
+            }
+            return con.preparedQuery("UPDATE " + titleEntriesTable(pool)
+                + " SET"
+                + " counterReportTitle = $2,"
+                + " printISSN = $4,"
+                + " onlineISSN = $5"
+                + " WHERE id = $1")
+                .execute(Tuple.of(id, counterReportTitle, printIssn, onlineIssn))
+                .map(id);
           }
-          final String match = onlineIssn;
-          return ermLookup(ctx, match).compose(erm -> {
+          return ermTitleLookup(ctx, onlineIssn).compose(erm -> {
             UUID kbTitleId = erm != null ? erm.getUUID(0) : null;
             String kbTitleName = erm != null ? erm.getString(1) : null;
             return con.preparedQuery("INSERT INTO " + titleEntriesTable(pool)
                 + "(id, counterReportTitle, matchCriteria,"
                 + " kbTitleName, kbTitleId,"
                 + " kbManualMatch, printISSN, onlineISSN)"
                 + " VALUES ($1, $2, $3, $4, $5, $6, $7, $8)"
-                + " ON CONFLICT (counterReportTitle) DO NOTHING")
-                .execute(Tuple.of(UUID.randomUUID(), counterReportTitle, onlineIssn,
+                + " ON CONFLICT (matchCriteria) DO NOTHING")
+                .execute(Tuple.of(UUID.randomUUID(), counterReportTitle, match,
                     kbTitleName, kbTitleId,
                     false, printIssn, onlineIssn))
                 .compose(x ->
                     con.preparedQuery("SELECT id FROM " + titleEntriesTable(pool)
-                        + " WHERE counterReportTitle = $1")
-                        .execute(Tuple.of(counterReportTitle))
+                        + " WHERE matchCriteria = $1")
+                        .execute(Tuple.of(match))
                         .map(res2 -> res2.iterator().next().getUUID(0))
                 );
           });
         });
   }
 
+  Future<Void> updateTitleFromAgreement(TenantPgPool pool, SqlConnection con,
+                                        UUID kbTitleId, RoutingContext ctx) {
+    if (kbTitleId == null) {
+      return Future.succeededFuture();
+    }
+    return con.preparedQuery("SELECT * FROM " + titleEntriesTable(pool)
+        + " WHERE kbTitleId = $1")
+        .execute(Tuple.of(kbTitleId))
+        .compose(res -> {
+          if (!res.iterator().hasNext()) {
+            return Future.succeededFuture();
+          }
+          return ermTitleLookup(ctx, kbTitleId).compose(erm -> {
```

Comment:
This deeply nested compose can be avoided:

```
          return ermTitleLookup(ctx, kbTitleId);
        }).compose(erm -> {
```

Reply:
Not really.